Hoist `jest.mock` with 2 arguments.

pvolok opened the pull request:
Comes from #796

How about throwing if second argument references outside variables? So we will be able to allow referencing later.

edit by @cpojer:
@bypass-lint internal eslint invariant-rule is silly

cpojer commented on packages/babel-plugin-jest-hoist/src/index.js:
```diff
@@ -10,9 +10,48 @@
+        const ids = [];
+        args[1].traverse(idVisitor, {ids});
+
+        for (var i = 0; i < ids.length; i++) {
```
should be `let`.

cpojer commented on packages/babel-plugin-jest-hoist/src/index.js:
```diff
@@ -10,9 +10,48 @@
+      if (args.length === 1) {
+        return args[0].isStringLiteral();
+      } else if (args.length === 2) {
+        const ids = [];
```
Should this be a `Set`?

cpojer commented on packages/babel-plugin-jest-hoist/src/index.js:
```diff
@@ -10,9 +10,48 @@
+          
+          while (scope !== args[1].scope) {
+            if (scope.bindings[id.node.name]) {
+              found = true;
```
this should just `return true`. That way we don't need the `found` variable at all.

thejameskyle replied:
You don't want to return, you want to break.

thejameskyle replied:
Sorry you don't want to break, you want to continue the outer loop

cpojer replied:
oh yes, you are right of course.

cpojer commented on packages/babel-plugin-jest-hoist/src/index.js:
```diff
@@ -10,9 +10,48 @@
+          
+          if (!found) {
+            throw new Error(
+              'The second argument of `jest.mock()` is not allowed to ' +
```
Let's prefix the message with `babel-jest-plugin-hoist: `.

cpojer commented:
Nice work! I left a couple of small comments. I do agree it should definitely throw when the second argument to mock is invalid and can't be hoisted.

I would like the API to be this:

```js
jest.mock('foo', () => {
  const myMock = require('myMock');
  myMock.foo = jest.fn();
  return myMock;
});
```

where the second argument to `jest.mock` is a function that gets run when `require('foo')` is called. I think this does require some minor changes to HasteModuleLoader – if you can make this work from the babel plugin side, then I can make the changes to the module loader with this new feature. It's fine if the integration test fails for now until I'm done with that :) Of course, if you'd like to take a stab at it, it would be somewhere around here: https://github.com/facebook/jest/blob/master/src/HasteModuleLoader/HasteModuleLoader.js#L159-L219 – and will require a new instance variable and some new tests. I think we can also change `_explicitlySetMocks` to this:

```
jest.setMock('foo', value) -> this._explicitlySetMocks[name] = () => value;

jest.mock('foo', () => value) -> this._explicitlySetMocks[name] = arg2OfJestMockCall;
```

which should make this change a bit easier. I'm not entirely sure about constraints around isolating require, but we can use this as a start.

We also need to update the documentation to include this new feature, this is something I can do.

Finally, this is a breaking API change and might confuse people. I think we need an escape-hatch for people to opt-out (just like we are keeping `jest.dontMock` around). We should add a function that is an alias for `jest.mock` but doesn't get hoisted. Since `jest.dontMock` is already quite silly, we can potentially just call this method `jest.doMock`. It's an escape-hatch that will be rarely used anyway :)

And also I wanna say: thank for the hard work you are putting into Jest. This is invaluable and I appreciate the support I'm receiving from you. Thank you.

pvolok commented:
I addressed the comments. Now it works this way: the plugin throws if second argument isn't function or if it references any outside argument except jest/require.

Thank you! I'm happy to have an opportunity to give something back to the community. Also you make the contributing experience really great!

cpojer commented on src/HasteModuleLoader/HasteModuleLoader.js:
```diff
@@ -159,8 +159,8 @@ class Loader {
-    if (moduleID in this._explicitlySetMocks) {
-      return this._explicitlySetMocks[moduleID];
+    if (moduleID in this._explicitlySetMockFactories) {
+      return this._explicitlySetMockFactories[moduleID]();
```
I think we need to make sure we only run this function once and that we cache the result. Otherwise we'll re-evaluate the mock every time it is being required. I can do that in a follow-up though :)

cpojer commented on packages/babel-plugin-jest-hoist/src/index.js:
```diff
@@ -8,11 +8,63 @@
+
+        const outerScope = args[1].parentPath.scope;
+
+        for (let id of ids) {
```
I think you can now use `const` here, actually.

pvolok replied:
In both node 5 and chrome `for (const a of [1, 2, 3]) console.log(a);` results in `1` being print 3 times.

hzoo replied:
^ prints 1, 2, 3 fine in chrome 49

cpojer replied:
yeah, works for me too?

pvolok replied:
Yesterday I had chrome 48, in 49 it works yes. But node still prints `1 1 1`. I'll add a comment so no one fix that.

cpojer commented on packages/babel-plugin-jest-hoist/src/index.js:
```diff
@@ -8,11 +8,63 @@
+
+          if (!found) {
+            invariant(
+              scope.hasGlobal(name) && (name === 'jest' || name === 'require'),
```
perfect. We should probably also allow `global`, what do you think? Let's make an object up there like

```js
const WHITELIST = {
  global: true,
  jest: true,
  process: true,
  require: true,
};
```

We'll probably have to add more things to this, like `Promise` and `setTimeout` etc. as well, but the list is exhaustive and it is ok with me to no allow everything there possibly can be.

pvolok replied:
Oh.. Can something like HTMLElement be here? If yes, maybe we should just allow all globals?

cpojer replied:
we don't know all globals upfront, but they can always be accessed through `global.<Foo>` so we have an escape hatch. The common ones should be whitelisted though.

pvolok replied:
What comes to mind right now: Object, Array, Map, WeakMap, Set, setInterval, Proxy, undefined, NaN, Infinity, RegExp, parseInt, parseFloat, and lots more. We will end up with a huge list that has to be maintained.

Plus I don't see how we improve user experience by checking all global variables. Let's review this example:

```js
import 'a';
import 'define-global-b';

jest.mock('a', () => b);
```

If we don't allow referencing global `b`, then the user will likely use `global.b` and might be confused why `b` is undefined. I would just add a note to the docs saying that mock factory will be executed when corresponding module is being imported.

What do you think?

cpojer commented:
Can you add a whitelist to the plugin instead? I can take care of the rest of the work in the HasteModuleLoader.

cpojer commented:
@facebook-github-bot import

cpojer commented:
@facebook-github-bot shipit

cpojer commented:
I decided to go with a whitelisting approach as follows:

* `jest` and `require` are whitelisted
* All current globals are whitelisted, these are the default globals from Node.js available in the Jest preprocessor which are most likely all non-custom globals that are also available inside of Jest's sandbox.
* All built-in types from the ES2015 standard are whitelisted.

This list is exhaustive and there is no reason we can't update it once per year when EcmaScript evolves. The escape-hatch is to access `global.Foo`. Now, this will not work if someone depends on mutations to built-in types if they occur after an import. However, these kind of mutations should be done in one of the `setupFiles`, so most likely this will be fine too.

I fixed a small bug around `setMock` that was introduced in this commit and updated the documentation.

The one thing I'm going back and forth on is whether the moduleFactory should receive it's own `require` implementation or whether it should just take the one from the parent context. It will likely slow things a little bit down but give more isolation to `jest.mock` module factories. Technically every module in Jest receives its completely separate require implementation. The API would look something like `jest.mock('Foo', require => require('mock'))`. I'm going to have to think about some of the downsides of doing this but it seems more technically correct. However, at the same time, the only thing that I think can really be affected by this is to copy the custom mock across test files, where the state of the module loader might be different.

Thanks again @pvolok for working on this.
